Aiming/Tracking/KF_tracker_modified.py

```diff
@@ -0,0 +1,253 @@
+
+    It stores the history of bounding boxes and ROIs, and can predict the
+    bounding box of the next frame.
+    """
+
+    def __init__(self, bbox, roi, frame_tick, armor_id):
+        """Initialize from prediction.
+
+        Args:
+            bbox (tuple): (center_x, center_y, w, h)
+            roi (np.ndarray): ROI of the armor
+            frame_tick (int): frame tick
+            armor_id (int): unique ID
+        """
+        self.bbox_buffer = [bbox]
+        self.roi_buffer = [roi]
+        self.observed_frame_tick = [frame_tick]
+        self.armor_id = armor_id  # unique ID
+        self.KF_matrix = KalmanTracker()
+        # self.KF_matrix = KalmanTracker(0.05) # dt = 0.05
+
+    def compute_cost(self, other_armor):
+        """Compute the cost of matching this armor with another armor.
+
+        Args:
+            other_armor (tracked_armor): another armor
+
+        Returns:
+            float: cost
+        """
+        assert isinstance(other_armor, tracked_armor)
+        # TODO: use more sophisticated metrics (e.g., RGB) as cost function
+        c_x, c_y, w, h = self.bbox_buffer[-1]
+        o_c_x, o_c_y, o_w, o_h = other_armor.bbox_buffer[-1]
+        return np.square(c_x - o_c_x) + np.square(c_y - o_c_y)
+
+    def update(self, other_armor, frame_tick):
+        """Update the state of this armor with matched armor.
+
+        Args:
+            other_armor (tracked_armor): another armor
+            frame_tick (int): frame tick
+        """
+        # Only call if these two armors are matched
+        self.bbox_buffer.append(other_armor.bbox_buffer[-1])
+        self.roi_buffer.append(other_armor.roi_buffer[-1])
+        self.observed_frame_tick.append(frame_tick)
+
+        # Maintain each armor's buffer so that anything older than
+        # FRAME_BUFFER_SIZE is dropped
+        self.bbox_buffer = self.bbox_buffer[-FRAME_BUFFER_SIZE:]
+        self.roi_buffer = self.roi_buffer[-FRAME_BUFFER_SIZE:]
+
+    def predict_bbox(self, cur_frame_tick):
+        """Predict the bounding box of the tracked armor at cur frame tick.
+
+        Args:
+            cur_frame_tick (int): current frame tick
+
+        TODO
+            - Use Kalman filter to do prediction
+            - Support future frame idx for predictions
+
+        Returns:
+            tuple: (center_x, center_y, w, h)
+        """
+        if cur_frame_tick == self.observed_frame_tick[-1] or len(
+                self.bbox_buffer) == 1:
+            # print(self.bbox_buffer[-1])
+            c_x, c_y, w, h = self.bbox_buffer[-1]
+            self.KF_matrix.update(c_x, c_y)
+            return self.bbox_buffer[-1]
+        else:
+            # KF tracking
+            c_x, c_y, w, h = self.bbox_buffer[-1]
+            predicted_x, predicted_y = self.KF_matrix.get_prediction()
+            self.KF_matrix.update(predicted_x, predicted_y)
+            return (int(predicted_x), int(predicted_y), w, h)
+
+
+class KF_tracker(object):
```

Comment:
This is redundant to the tracker code in the basic tracker. Can we merge these two files and support specifying filtering method using config.py?